flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileMonitoringFunction.java

```diff
@@ -93,20 +102,45 @@
 	/** Which new data to process (see {@link FileProcessingMode}. */
 	private final FileProcessingMode watchType;
 
-	/** The maximum file modification time seen so far. */
+	/** The offset interval back from the latest file modification timestamp to scan for our-of-order files.
+	 *  Valid value for this is from 0 to Long.MAX_VALUE.
+	 *
+	 *  <p><b>NOTE: </b>: Files with (modTime > Long.MIN_VALUE + readConsistencyOffset) will NOT be read.
+	 */
+	private final long readConsistencyOffset;
+
+	/** The current modification time watermark. */
 	private volatile long globalModificationTime = Long.MIN_VALUE;
```

Comment:
Is it possible that we could not depend on this variable? And I think this would reduce the complexity of restoring logical.

Reply:
As I explained in the corresponding JIRA ticket https://issues.apache.org/jira/browse/FLINK-9940, yes, this is needed.